src/utils/mod.rs

```diff
@@ -1516,3 +1517,47 @@ impl From<PcrData> for TPML_DIGEST {
         tpml_digest
     }
 }
+
+fn int_to_string(num: u32) -> String {
+    let mut num = num;
+    let mut result = String::new();
+
+    loop {
+        let chr: u8 = (num & 0xFF) as u8;
+        num >>= 8;
+        if chr != 0 {
+            let chr = char::from(chr);
+            result.insert(0, chr);
+        }
+        if num == 0 {
+            break;
+        }
+    }
+    result
+}
```

Comment:
I think this is covered by [`ToString`](https://doc.rust-lang.org/std/primitive.u32.html#impl-ToString) (?)

Reply:
Are you sure?

**TCG TPM2 r1p59 Part 2 Structures**
>|Capability Name| Value | Comments|
>|-------------|-------------|-------------|
>|TPM_PT_VENDOR_STRING_1| PT_FIXED + 6| <p>the first four characters of the vendor ID string<br>NOTE: <br>When the vendor string is fewer than 16 octets, the additional property values do not have to be present. A vendor string of 4 octets can be represented in one 32-bit value and no null terminating character is required.</p>|
>|TPM_PT_VENDOR_STRING_2|PT_FIXED + 7|the second four characters of the vendor ID string|
>|TPM_PT_VENDOR_STRING_3|PT_FIXED + 8|the third four characters of the vendor ID string|
>|TPM_PT_VENDOR_STRING_4|PT_FIXED + 9|the fourth four characters of the vendor ID sting|

Reply:
Hmm, having a 2nd look the function I linked is definitely not what was intended, probably [this](https://doc.rust-lang.org/std/str/fn.from_utf8.html) is more appropriate. I think the section you quote is more relevant to the function below, that actually handles the full vendor string - this seems to be intended mostly as a `u32` -> `String` conversion, by treating the 4 octets as chars.

Reply:
I do not like that loop. It should be ensured to only use the bytes in the value.

I would must more prefer it looking something like this (OBS psedo code)
```rust
let mut result = String::new();
for byte  in num.as_be_bytes() {
    if byte == 0 {
       break;
    }
    result.push(char::from(byte));
}
result
```

Reply:
Yep, I just wrote some code like that last version an hour ago, but was in calls so couldn't submit. Will do that shortly :).

Reply:
> Hmm, having a 2nd look the function I linked is definitely not what was intended, probably [this](https://doc.rust-lang.org/std/str/fn.from_utf8.html) is more appropriate. I think the section you quote is more relevant to the function below, that actually handles the full vendor string - this seems to be intended mostly as a `u32` -> `String` conversion, by treating the 4 octets as chars.

Yes, the interesting part there was the note section. I am not sure the from_utf8 would work either because the specification says the string is in ASCII (but I might be wrong?).

Reply:
I think `from_utf8` should work - ASCII is a subset of UTF8. What I understand from that comment on `TPM_PT_VENDOR_STRING_1` is that the other vendor string values might be 0, in which case they shouldn't be included in the string, but those first 4 octets should be non-null. Maybe? I'm not sure

Reply:
Yeah, the null values are already prevented by 
```
    // Filter out the empty values
    .filter_map(|x| *x)
    // Filter out zero values
    .filter(|x| *x != 0)
```

Reply:
But I think I'm going with @Superhepper 's approach, given that interpreting it as UTF8 while it's really just individual ASCII characters seems not the best.

Reply:
I'm ok with that approach, though [`char::from(u8)`](https://doc.rust-lang.org/std/primitive.char.html#impl-From%3Cu8%3E) isn't really ASCII either, it's just compatible with.